File: Solution MaxHeap.java

import java.util.PriorityQueue;

public class Solution {

    private record Element(int value, int index) {}

    public int[] maxSlidingWindow(int[] inputNumbers, int windowSize) {

        PriorityQueue<Element> maxHeap = new PriorityQueue<>((first, second) -> second.value - first.value);
        for (int i = 0; i < windowSize; ++i) {
            maxHeap.add(new Element(inputNumbers[i], i));
        }

        int[] maxValuesWindow = new int[inputNumbers.length - windowSize + 1];
        maxValuesWindow[0] = maxHeap.peek().value;

        for (int i = windowSize; i < inputNumbers.length; ++i) {
            while (!maxHeap.isEmpty() && maxHeap.peek().index < i - windowSize + 1) {
                maxHeap.poll();
            }
            maxHeap.add(new Element(inputNumbers[i], i));
            maxValuesWindow[i - windowSize + 1] = maxHeap.peek().value;
        }

        return maxValuesWindow;
    }
}

File: Solution MaxHeap.cpp

#include <queue>
#include <vector>
using namespace std;

class Solution {

    struct Element {
        int value{};
        int index{};
        Element(int value, int index) : value{value}, index{index}{}
    };
    
    inline static auto maxValueOnTop = [](const Element& first, const Element& second) {
        return first.value < second.value;
    };

public:
    vector<int> maxSlidingWindow(vector<int>& inputNumbers, int windowSize) {
        
        priority_queue<Element, vector<Element>, decltype(maxValueOnTop) > maxHeap(maxValueOnTop);
        for (int i = 0; i < windowSize; ++i) {
            maxHeap.emplace(Element(inputNumbers[i], i));
        }

        vector<int> maxValuesWindow(inputNumbers.size() - windowSize + 1);
        maxValuesWindow[0] = maxHeap.top().value;

        for (int i = windowSize; i < inputNumbers.size(); ++i) {
            while (!maxHeap.empty() && maxHeap.top().index < i - windowSize + 1) {
                maxHeap.pop();
            }
            maxHeap.emplace(Element(inputNumbers[i], i));
            maxValuesWindow[i - windowSize + 1] = maxHeap.top().value;
        }

        return maxValuesWindow;
    }
};
